crates/relayer/src/chain/ckb4ibc/utils.rs

```diff
@@ -467,3 +511,101 @@ pub fn transaction_to_event(
     };
     Ok(event)
 }
+
+pub fn get_ibc_merkle_proof(height: Height, encoded: Vec<u8>) -> Result<Proofs, Error> {
+    let useless_client_proof = vec![0u8].try_into().unwrap();
+    let useless_consensus_proof =
+        ConsensusProof::new(vec![0u8].try_into().unwrap(), Height::default()).unwrap();
+    Proofs::new(
+        encoded.try_into().unwrap(),
+        Some(useless_client_proof),
+        Some(useless_consensus_proof),
+        None,
+        height,
+    )
+    .map_err(|err| Error::other_error(err.to_string()))
+}
+
+#[derive(EthAbiCodec, EthAbiType)]
+struct AxonObjectProof {
+    pub ckb_transaction: Vec<u8>,
+    pub block_hash: [u8; 32],
+    pub proof_payload: VerifyProofPayload,
+}
+
+pub async fn generate_tx_proof_from_block(
+    rpc_client: &impl CkbReader,
+    tx_hash: &H256,
+) -> Result<Option<Proofs>, Error> {
+    let block_hash = rpc_client
+        .get_transaction(tx_hash)
+        .await?
+        .map(|v| v.tx_status.block_hash);
+    let Some(Some(block_hash)) = block_hash else {
+        return Err(Error::other_error(format!(
+            "cannot find block_hash from tx {}",
+            hex::encode(tx_hash)
+        )));
+    };
+
+    // collect transaction hashes from block
+    let mut transaction: Option<Transaction> = None;
+    let block = rpc_client.get_block(&block_hash).await?;
+    let tx_hashes = block
+        .transactions
+        .iter()
+        .map(|tx| {
+            if &tx.hash == tx_hash {
+                transaction = Some(tx.inner.clone().into());
+            }
+            tx.hash.clone()
+        })
+        .collect_vec();
+    let witness_hashes = block
+        .transactions
+        .into_iter()
+        .map(|tx| Transaction::from(tx.inner).calc_witness_hash().unpack())
+        .collect_vec();
+
+    let Some(transaction) = transaction else {
+        return Ok(None);
+    };
+
+    // generate transaction proof
+    let TransactionAndWitnessProof {
+        block_hash,
+        transactions_proof: _,
+        witnesses_proof: proof,
+    } = rpc_client
+        .get_transaction_and_witness_proof(tx_hashes.clone(), block.header.hash)
+        .await?;
+
+    let raw_transaction_root = merkle_root(&tx_hashes.iter().map(Pack::pack).collect_vec());
+    let witnesses_root = merkle_root(&witness_hashes.iter().map(Pack::pack).collect_vec());
+
+    let proof_payload = VerifyProofPayload {
+        verify_type: 1, // to verify witness
+        transactions_root: block.header.inner.transactions_root.into(),
+        witnesses_root: witnesses_root.unpack().into(),
+        raw_transactions_root: raw_transaction_root.unpack().into(),
+        proof: Proof {
+            indices: proof.indices.into_iter().map(Into::into).collect(),
+            lemmas: proof.lemmas.into_iter().map(Into::into).collect(),
+            leaves: witness_hashes.into_iter().map(Into::into).collect(),
```

Comment:
It's strange that all witnesses hashes are put into the proof as leaves.

Reply:
check again